### ISSUE TITLE
Azure AD authentication issue with trusted locations

### ISSUE THREAD

**Issue by CorHortensius** [snok/django-auth-adfs#67]
We managed to integrate django_auth_adfs in our system. And partly it's also working with Azure AD.
but sometimes we got the following error:
ADFS server returned an error: AADSTS50076: Due to a configuration change made by your administrator, or because you moved to a new location, you must use multi-factor authentication to access `'-*****--*-0c4812*****' `.

Azure AD is configured with conditional access rules. When you are in a trusted location you don't have to enter 2FA, but if you are not on a trusted location you have to enter 2FA.
When you are on a NOT trusted location, you can login fine.
When you are on a trusted location we got the error above.

```
ADFS server returned an error: AADSTS50076: Due to a configuration change made by your 
administrator, or because you moved to a new location, you must use multi-factor authentication to 
access '*******-*****-****-****-0c4812*****'.
```

Things we tried:
- disable trusted location. (Then you can login)
- exclude the user from conditional access rule. (then you can login)
- disable 2 factor authentication for the user. (then you can login)
- enable trusted location (you can not login anymore)

We found an article from Microsoft that explains exactly what happens:

https://docs.microsoft.com/en-us/azure/active-directory/develop/conditional-access-dev-guide#scenario-app-performing-the-on-behalf-of-flow

**Comment by JonasKs** [comment]
Seems to have been reported in #62 too, after it was closed.

**Comment by jobec** [comment]
I'll try to replicate it.

Did make any other changes to you Azure AD settings, apart from those described here?
https://django-auth-adfs.readthedocs.io/en/latest/azure_ad_config_guide.html
so that I'm sure I'm simulating the right thing.

**Comment by jobec** [comment]
I tried to replicate this, but given the fact that I use a free Azure AD account, I can't set those kind of "Conditional Access" policies.

So I'm afraid I can't test or fix it 😞 

If anyone can show me an example for the expected flow in HTTP requests with the entire request and response body, I can try to implement it. Pull requests are also always welcome!

**Comment by CorHortensius** [comment]
@jobec I will try this week to get the http request flow from azure.

**Comment by jobec** [comment]
@CorHortensius Did you find time already?

**Comment by CorHortensius** [comment]
@jobec I'm setting up a shadow environment to test this. It's almost done.
can you point me to, what the best place is to put print statements in the code so we can see what is happening ?

**Comment by jobec** [comment]
I would say turn on debug logging in Django:
https://django-auth-adfs.readthedocs.io/en/latest/troubleshooting.html

This will show where things go wrong.

And depending on where it goes wrong, you'll need to add some debug logs to dump ADFS responses.

But I would rather approach this by simulating the call flow in something like Postman.
1. Request from the client to ADFS after it was redirected by Django.
1. Response from ADFS with a redirect and the authorization code.
1. Request from Django to ADFS to exchange the authorization code into an access token.

I'm guessing step 3 will be failing due to the AADSTS50076 error response.

Next step is figuring our what comes after step 3.
* Should the client be redirected to somewhere specific?
* Should Django add some special parameters to step 3?
* Should Django request the token at some other location?
* Should there be some settings changed on the Azure side?

**Comment by CorHortensius** [comment]
oke, Got the first terminal debug output.

```
March 28, 2019 - 09:13:40
Django version 1.11, using settings '.settings'
Starting development server at http://0:8005/
Quit the server with CONTROL-C.

[28/Mar/2019 09:14:04] "GET /logout/?next=/ HTTP/1.0" 302 0
[28/Mar/2019 09:14:04] "GET / HTTP/1.0" 302 0
The request's session was deleted before the request completed. The user may have logged out in a concurrent request, for example.
DEBUG 2019-03-28 09:14:04,267 django_auth_adfs Loading django_auth_adfs ID Provider configuration.
INFO 2019-03-28 09:14:04,267 django_auth_adfs Trying to get OpenID Connect config from https://login.microsoftonline.com/**cfa***-****-****-****-700*********/.well-known/openid-configuration
[28/Mar/2019 09:14:04] "GET / HTTP/1.0" 400 20340
DEBUG 2019-03-28 09:14:04,416 django_auth_adfs Loading public key from certificate: publickey..............................................................................................
DEBUG 2019-03-28 09:14:04,416 django_auth_adfs Loading public key from certificate: publickey..............................................................................................
DEBUG 2019-03-28 09:14:04,417 django_auth_adfs Loading public key from certificate: publickey..............................................................................................
INFO 2019-03-28 09:14:04,417 django_auth_adfs django_auth_adfs loaded settings from ADFS server.
INFO 2019-03-28 09:14:04,417 django_auth_adfs operating mode:         openid_connect
INFO 2019-03-28 09:14:04,417 django_auth_adfs authorization endpoint: https://login.microsoftonline.com/**cfa***-****-****-****-700*********/oauth2/authorize
INFO 2019-03-28 09:14:04,417 django_auth_adfs token endpoint:         https://login.microsoftonline.com/**cfa***-****-****-****-700*********/oauth2/token
INFO 2019-03-28 09:14:04,417 django_auth_adfs end session endpoint:   https://login.microsoftonline.com/**cfa***-****-****-****-700*********/oauth2/logout
INFO 2019-03-28 09:14:04,417 django_auth_adfs issuer:                 https://sts.windows.net/**cfa***-****-****-****-700*********/


###################### Try from Untrusted location
[28/Mar/2019 09:15:17] "GET / HTTP/1.0" 302 0
[28/Mar/2019 09:15:17] "GET /oauth2/login?next=/ HTTP/1.0" 302 0
DEBUG 2019-03-28 09:15:46,730 django_auth_adfs Received authorization code: AQABAAIAAACEfexX.........................................................lDWjojIAA
DEBUG 2019-03-28 09:15:46,730 django_auth_adfs Getting access token at: https://login.microsoftonline.com/**cfa***-****-****-****-700*********/oauth2/token
DEBUG 2019-03-28 09:15:46,944 django_auth_adfs Received access token: eyJ0eXAiOiJKV1Q..................................................Vn2yiRv7W8vAhyl24_Q
WARNING 2019-03-28 09:15:46,956 django_auth_adfs Claim 'given_name' for user field 'first_name' was not found in the access token for user 'mail@mailDomain.com'. Field is not required and will be left empty
WARNING 2019-03-28 09:15:46,957 django_auth_adfs Claim 'family_name' for user field 'last_name' was not found in the access token for user 'mail@mailDomain.com'. Field is not required and will be left empty
WARNING 2019-03-28 09:15:46,957 django_auth_adfs Claim 'email' for user field 'email' was not found in the access token for user 'mail@mailDomain.com'. Field is not required and will be left empty
DEBUG 2019-03-28 09:15:46,959 django_auth_adfs The configured groups claim 'groups' was not found in the access token
DEBUG 2019-03-28 09:15:46,959 django_auth_adfs The configured group claim was not found in the access token
[28/Mar/2019 09:15:46] "GET /oauth2/callback?code=AQABAAIAAACEfexXxjamQb3...........state=Lw%3d%3d&session_state=8888888-c4888838-8888-88888-888888 HTTP/1.0" 302 0
[28/Mar/2019 09:15:47] "GET / HTTP/1.0" 200 8729
[28/Mar/2019 09:15:47] "GET /dashboard/ HTTP/1.0" 200 1139
[28/Mar/2019 09:15:47] "GET /portal/CORP/main HTTP/1.0" 200 371

####### The above try is working from an Untrusted location

####### Now we are gone g try from a Trusted Location
[28/Mar/2019 09:17:45] "GET / HTTP/1.0" 302 0
[28/Mar/2019 09:17:45] "GET /oauth2/login?next=/ HTTP/1.0" 302 0
DEBUG 2019-03-28 09:17:59,711 django_auth_adfs Received authorization code: ........................................................................................................
DEBUG 2019-03-28 09:17:59,711 django_auth_adfs Getting access token at: https://login.microsoftonline.com/**cfa***-****-****-****-700*********/oauth2/token
ERROR 2019-03-28 09:17:59,991 django_auth_adfs ADFS server returned an error: AADSTS50076: Due to a configuration change made by your administrator, or because you moved to a new location, you must use multi-factor authentication to access 'a88*****-****-****-****-74*********7'.

Trace ID: 85*****-****-****-****-47********00

Correlation ID: 58*****-****-****-****-9b*********

Timestamp: 2019-03-28 08:17:59Z
Internal Server Error: /oauth2/callback
Traceback (most recent call last):
  File "/home/username/test_APP/lib/python3.6/site-packages/django/core/handlers/exception.py", line 41, in inner
    response = get_response(request)
  File "/home/username/test_APP/lib/python3.6/site-packages/django/core/handlers/base.py", line 187, in _get_response
    response = self.process_exception_by_middleware(e, request)
  File "/home/username/test_APP/lib/python3.6/site-packages/django/core/handlers/base.py", line 185, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/home/username/test_APP/lib/python3.6/site-packages/django/views/generic/base.py", line 68, in view
    return self.dispatch(request, *args, **kwargs)
  File "/home/username/test_APP/lib/python3.6/site-packages/django/views/generic/base.py", line 88, in dispatch
    return handler(request, *args, **kwargs)
  File "/home/username/test_APP/lib/python3.6/site-packages/django_auth_adfs/views.py", line 62, in get
    }, status=401)
  File "/home/username/test_APP/lib/python3.6/site-packages/django/shortcuts.py", line 30, in render
    content = loader.render_to_string(template_name, context, request, using=using)
  File "/home/username/test_APP/lib/python3.6/site-packages/django/template/loader.py", line 67, in render_to_string
    template = get_template(template_name, using=using)
  File "/home/username/test_APP/lib/python3.6/site-packages/django/template/loader.py", line 25, in get_template
    raise TemplateDoesNotExist(template_name, chain=chain)
django.template.exceptions.TemplateDoesNotExist: django_auth_adfs/login_failed.html
[28/Mar/2019 09:18:00] "GET /oauth2/callback?code=AQABAAIAAAC........H-jtoHCAA&state=Lw%3d%3d&session_state=00000000-0000-0000-0000-0000000000 HTTP/1.0" 500 114184

############  Failed to login from a trusted location
```

**Comment by jobec** [comment]
Try changing this:
https://github.com/jobec/django-auth-adfs/blob/07197be392724d16a6132b03d9eafb1d634749cf/django_auth_adfs/backend.py#L32-L34
into:
```python
        if response.status_code == 400:
            logger.error("ADFS server returned an error: " + repr(response.json()))
            raise PermissionDenied
```
And run it again.

**Comment by jobec** [comment]
It should dump the entire response from ADFS including an error and claims field.

**Comment by CorHortensius** [comment]
azure error codes:
ref: https://docs.microsoft.com/en-us/azure/active-directory/reports-monitoring/reference-sign-ins-error-codes
50076 | User did not pass the MFA challenge (non interactive).

https://social.msdn.microsoft.com/Forums/azure/en-US/9ddd0129-9355-4d68-a413-7847e3da2d03/what-is-50076user-did-not-pass-the-mfa-challenge-non-interactive?forum=windowsazureactiveauthentication

```
[28/Mar/2019 10:57:12] "GET /oauth2/login?next=/ HTTP/1.0" 302 0
DEBUG 2019-03-28 10:57:19,555 django_auth_adfs Received authorization code: AQABAAIAAACEf.....iAA
DEBUG 2019-03-28 10:57:19,555 django_auth_adfs Getting access token at: https://login.microsoftonline.com/39cfa0e4-669a-4611-9363-700b0bcc9054/oauth2/token
ERROR 2019-03-28 10:57:19,775 django_auth_adfs ADFS server returned an error: {'error': 'interaction_required', 'error_description': "AADSTS50076: Due to a configuration change made by your administrator, or because you moved to a new location, you must use multi-factor authentication to access '********-****-****-****-**********'.\r\nTrace ID: ********-****-****-****-**********\r\nCorrelation ID: ********-****-****-****-**********\r\nTimestamp: 2019-03-28 09:57:19Z", 'error_codes': [50076], 'timestamp': '2019-03-28 09:57:19Z', 'trace_id': '********-****-****-****-**********', 'correlation_id': '********-****-****-****-**********', 'suberror': 'basic_action'}


Internal Server Error: /oauth2/callback
Traceback (most recent call last):
  File "/home/chortensius/taxmonitor_SSO/lib/python3.6/site-packages/django/core/handlers/exception.py", line 41, in inner
    response = get_response(request)
  File "/home/chortensius/taxmonitor_SSO/lib/python3.6/site-packages/django/core/handlers/base.py", line 187, in _get_response
    response = self.process_exception_by_middleware(e, request)
  File "/home/chortensius/taxmonitor_SSO/lib/python3.6/site-packages/django/core/handlers/base.py", line 185, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/home/chortensius/taxmonitor_SSO/lib/python3.6/site-packages/django/views/generic/base.py", line 68, in view
    return self.dispatch(request, *args, **kwargs)
  File "/home/chortensius/taxmonitor_SSO/lib/python3.6/site-packages/django/views/generic/base.py", line 88, in dispatch
    return handler(request, *args, **kwargs)
  File "/home/chortensius/taxmonitor_SSO/lib/python3.6/site-packages/django_auth_adfs/views.py", line 62, in get
    }, status=401)
  File "/home/chortensius/taxmonitor_SSO/lib/python3.6/site-packages/django/shortcuts.py", line 30, in render
    content = loader.render_to_string(template_name, context, request, using=using)
  File "/home/chortensius/taxmonitor_SSO/lib/python3.6/site-packages/django/template/loader.py", line 67, in render_to_string
    template = get_template(template_name, using=using)
  File "/home/chortensius/taxmonitor_SSO/lib/python3.6/site-packages/django/template/loader.py", line 25, in get_template
    raise TemplateDoesNotExist(template_name, chain=chain)
django.template.exceptions.TemplateDoesNotExist: django_auth_adfs/login_failed.html
[28/Mar/2019 10:57:20] "GET /oauth2/callback?code=AQABAAIAAACEf.................iAA&state=Lw%3d%3d&session_state=d9*****9-****-****-****-11****30 HTTP/1.0" 500 114186
```

**Comment by jobec** [comment]
Microsoft and it's documentation... 🤢 

Can you try settings `DISABLE_SSO` to `True`?
https://django-auth-adfs.readthedocs.io/en/latest/settings_ref.html#disable-sso
Does that change anything?

**Comment by CorHortensius** [comment]
Nope same issue.

But I think I know where the problem sits.
I added my external server IP to the trusted locations, and then the login process is performed right.
When I remove the external server IP from the trusted locations it's not working anymore.

Client | Appserver | Passing
-- | -- | --
Trusted | Trusted | Yes
Trusted | Untrusted | No
Untrusted | Untrusted | Yes

**Comment by jobec** [comment]
What do you mean with "trusted locations" ? In Azure?
In your table, what's the result from "untrusted", "trusted" ?

The big issue here is, that I can't find any documentation about what to do when you get that error. It says to restart the authentication interactively and send received claims along. But there's zero docs about how exactly to do that.

**Comment by CorHortensius** [comment]
@jobec I updated my comment.

**Comment by JonasKs** [comment]
@CorHortensius, I made a PoC today on how this can be solved, see #101 .

In your [link](This new request requires the user to do multi-factor authentication and then send this new token back to Web API 1 and complete the on-behalf-of flow.) it says:  
> This new request requires the user to do multi-factor authentication and then send this new token back to Web API 1 and complete the on-behalf-of flow.

The way I read this, is that it's required for the user to do a completely new authentication, this time with MFA. I've provided some explanation on the implementation in the pull request and how you can install it to test it out.

**Comment by JonasKs** [comment]
Fix released in `1.4.1`. It's published to PyPI.
